### ISSUE TITLE
GraphQL nested location devices query does not work

### ISSUE THREAD

**Issue by zsblevins** [nautobot/nautobot#6535]
### Environment
* Nautobot version (Docker tag too if applicable): 2.3.8
* Python version: 3.11
* Database platform, version: Postgres
* Middleware(s): None

### Steps to Reproduce
1. Create tag "route-server" applicable to dcim|device
2. Create Location Type Site
3. Create Location Type Module underneath Site
4. Create Site "SiteA"
5. Create Module "Module 1" under Site "SiteA"
6. Create Module "Module 2" under Site "SiteA"
7. Create device "DeviceA" in Module "Module 1"
8. Tag DeviceA with "route-server"
9. Create device "DeviceB" in Module "Module 2"
10. Perform the following GraphQL query with `{"id": <DeviceB_UUID>}`:

```
fragment remote_peer on DeviceType {
  name
}

query ($id: ID!) {
  devices(tags: ["route-server"], location: "SiteA") {
    ...remote_peer
  }
  device(id: $id) {
    location {
      name
      parent {
        name
        devices(tags: ["route-server"]) {
          ...remote_peer
        }
      }
    }
  }
}
```

### Expected Behavior
The top-level devices should match the location -> parent -> devices, e.g.

```
{
  "data": {
    "devices": [
      {
        "name": "DeviceA",
      }
    ],
    "device": {
      "location": {
        "name": "Module 2",
        "parent": {
          "name": "SiteA",
          "devices": [
            {
              "name": "DeviceA",
            }
          ]
        }
      }
    }
  }
}
```

### Observed Behavior
The devices do not appear under the location parent object:
```
{
  "data": {
    "devices": [
      {
        "name": "DeviceA",
      }
    ],
    "device": {
      "location": {
        "name": "Module 2",
        "parent": {
          "name": "SiteA",
          "devices": []
        }
      }
    }
  }
}
```

**Comment by zsblevins** [comment]
Because this query is done from the perspective of DeviceB, which exists in Module 2.

**Comment by zsblevins** [comment]
Any updates here?

**Comment by cmacnevin** [comment]
This is a blocking bug for us, is it being looked into?

Thanks

**Comment by glennmatthews** [comment]
Sorry, looking at this issue a bit later - is the issue that you're expecting the `devices`reported for `SiteA` to also include devices belonging to its descendant locations, e.g. devices assigned to `Module 2`? That's not how the data model currently operates.

**Comment by zsblevins** [comment]
That's correct, it works if you query the location -> devices at the top level, as demonstrated in the output I shared, but not if you go device -> location -> parent -> devices.

I think the modeling is there, it's just not consistently applied.

**Comment by glennmatthews** [comment]
Filtering logic versus data lookup logic, yeah. Thanks for the clarification.

**Comment by zsblevins** [comment]
Assuming our nesting levels stay consistent, I can still get all the data I need for a given device in a single query by doing:
```
fragment remote_peer on DeviceType {
  name
}

query ($id: ID!) {
  device(id: $id) {
    location {
      name
      devices(tags: ["route-server"]) {
        ...remote_peer
      }
      parent {
        name
        children{
          devices(tags:["route-server"]){
            ...remote_peer
          }
        }
      }
    }
  }
}
```

But we have some variance today in how things are modeled across different Sites, so this does cause us to bloat our queries quite a bit since we need to repeat the `devices` query at each location layer that has the potential to be a Site. For context, yhis is a subset of a query done to pull _all_ configuration data for a given Device in a single query by Device ID, where device UUID is the only input. If this behavior can be adjusted, it'll allow us to greatly simplify our querying and reduce bloat in the response as well.

**Comment by glennmatthews** [comment]
Related: #7146, #7458
